File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part02_gridtraveller/java/GridTraveller.java

public class GridTraveller {
  public static void main(String[] args) {
    System.out.println(travelGrid(1,1));
    System.out.println(travelGrid(2,3));
    System.out.println(travelGrid(3,2));
    System.out.println(travelGrid(3,3));
    System.out.println(travelGrid(10,12));
  } 
  
  public static long travelGrid(int m, int n){
    if(m == 1 && n == 1){return 1;}
    if(m == 0 || n == 0){return 0;}

    return travelGrid(m-1, n) + travelGrid(m, n-1);
  }
}

File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part07_tabulation/02_gridtraveller/javascript/gridTraveller.js
const gridTraveller = (m, n) => {
  const table = Array(m + 1)
    .fill()
    .map(() => Array(n + 1).fill(0));

  table[1][1] = 1;

  for (let i = 0; i <= m; i++) {
    for (let j = 0; j <= n; j++) {
      const current = table[i][j];
      if (j + 1 <= n) table[i][j + 1] += current;
      if (i + 1 <= m) table[i + 1][j] += current;
    }
  }
  //console.log(table);
  return table[m][n];
};

console.log(gridTraveller(1, 1)); // 1
console.log(gridTraveller(2, 3)); // 3
console.log(gridTraveller(3, 2)); // 3
console.log(gridTraveller(3, 3)); // 6
console.log(gridTraveller(18, 18)); // 2333606220
